File: algo/112. Path Sum.js
// @9 Divide and Conquer

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {boolean}
 */

var hasPathSum = function(root, sum) {
    
    // if it is the child of leaf, return false
    if (root == null) {
        return false;
    }
    
    // if it is the leaf, check the value
    if (root.left==null && root.right==null) {
        if (root.val == sum) {
            return true;
        } else {
            return false;
        }
    }
    
    // split
    var leftRes = hasPathSum(root.left, sum-root.val);
    var rightRes = hasPathSum(root.right, sum-root.val);
    
    // merge
    if (leftRes || rightRes) {
        return true;
    }
    
    return false;
};

File: round3/112. Path Sum.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    public bool HasPathSum(TreeNode root, int targetSum) {
        if (root == null) {
            return false;
        }
        if (root.left == null && root.right == null) {
            if (root.val == targetSum) return true;
            return false;
        }
        
        bool leftRes = HasPathSum(root.left, targetSum - root.val);
        bool rightRes = HasPathSum(root.right, targetSum - root.val);
        
        if(leftRes || rightRes) return true;
        return false;
    }
}